File: tscode/bitseal_rtc/e2e.test.ts
/** @vitest-environment node */
// @ts-nocheck
import { Fragmenter, Reassembler } from './Fragment'
import { Session } from './BitSealRTC.js'
import PrivateKey from '@bsv/sdk/primitives/PrivateKey'
import { randomBytes } from 'crypto'
import { describe, test, expect } from 'vitest'

function priv(byteVal:number):PrivateKey {
  const bytes = new Array(32).fill(0)
  bytes[31] = byteVal
  return new PrivateKey(bytes)
}

describe('BitSeal-RTC e2e', () => {
  test('out-of-order & duplicate reassembly (4MiB)', () => {
    const alice = priv(1)
    const bob = priv(2)
    const saltA = [1,2,3,4]
    const saltB = [5,6,7,8]

    const sessA = Session.create(alice, bob.toPublicKey(), saltA, saltB)
    const sessB = Session.create(bob, alice.toPublicKey(), saltB, saltA)

    const fragA = new Fragmenter(sessA)
    const recvB = new Reassembler(sessB)

    const msg = Array.from(randomBytes(4<<20)) // 4 MiB
    const frames:number[][] = fragA.encode(msg).map(f=>Array.from(f))

    // duplicate 10 random frames (append after originals to keep order relatively intact)
    for(let i=0;i<10;i++){
      const idx = Math.floor(Math.random()*frames.length)
      frames.push(frames[idx])
    }
    // shuffle within windows of 32 to stay within replay window (64)
    const chunk=32
    for(let start=0;start<frames.length;start+=chunk){
      const end=Math.min(start+chunk,frames.length)
      for(let i=start;i<end-1;i++){
        const j = start + Math.floor(Math.random()*(end-start))
        const tmp=frames[i];frames[i]=frames[j];frames[j]=tmp
      }
    }

    let assembled:Uint8Array|null=null
    for(const f of frames){
      try{
        const res=recvB.push(Uint8Array.from(f))
        if(res.done) assembled=res.msg!
      }catch{ /* ignore replay/decrypt fails */ }
    }
    expect(assembled).not.toBeNull()
    expect(Array.from(assembled!)).toStrictEqual(msg)
  })

  test('replay window boundary rejects old seq', () => {
    const key = priv(3)
    const salt=[9,9,9,9]
    const sess = Session.create(key, key.toPublicKey(), salt, salt)

    const frames:Uint8Array[]=[]
    for(let i=0;i<70;i++){
      frames.push(Uint8Array.from(sess.encode([i])))
    }
    // decode first 70 sequentially
    frames.forEach(f=>sess.decode(Array.from(f)))

    // replay earliest frame (seq distance >64)
    expect(()=>sess.decode(Array.from(frames[0]))).toThrow()
  })
}) 

File: gocode/bitseal_rtc/e2e_test.go
package rtc

import (
	"bytes"
	crand "crypto/rand"
	"encoding/binary"
	"math/rand"
	"testing"
	"time"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	"go.uber.org/zap/zaptest"
)

func mustPriv(seed byte) *ec.PrivateKey {
	b := bytes.Repeat([]byte{seed}, 32)
	key, _ := ec.PrivateKeyFromBytes(b)
	return key
}

// TestOutOfOrderDuplicate checks reassembler under out-of-order & duplicate delivery.
func TestOutOfOrderDuplicate(t *testing.T) {
	rand.Seed(time.Now().UnixNano())

	privA := mustPriv(0x01)
	privB := mustPriv(0x02)
	saltA := []byte{1, 2, 3, 4}
	saltB := []byte{5, 6, 7, 8}

	logger := zaptest.NewLogger(t)
	sessA, _ := NewSession(privA, privB.PubKey(), saltA, saltB, logger)
	sessB, _ := NewSession(privB, privA.PubKey(), saltB, saltA, logger)

	fragA := NewFragmenter(sessA)
	recvB := NewReassembler(sessB)

	// 4 MiB random message
	msg := make([]byte, 4<<20)
	_, _ = crand.Read(msg)

	frames, err := fragA.Encode(msg)
	if err != nil {
		t.Fatal(err)
	}

	// introduce duplicates (kept after originals to stay within window)
	for i := 0; i < 10 && len(frames) > 0; i++ {
		idx := rand.Intn(len(frames))
		frames = append(frames, frames[idx])
	}

	var assembled []byte
	for _, f := range frames {
		plain, ok, err := recvB.Push(f)
		if err != nil {
			// duplicate or replay frames may error; skip them
			continue
		}
		if ok {
			assembled = plain
		}
	}
	if assembled == nil {
		t.Fatal("failed to assemble message; not fully reassembled")
	}
	if !bytes.Equal(assembled, msg) {
		t.Fatal("assembled payload mismatch")
	}
}

// TestReplayWindowBoundary ensures packets older than window are rejected.
func TestReplayWindowBoundary(t *testing.T) {
	priv := mustPriv(0x03)
	salt := []byte{1, 1, 1, 1}
	logger := zaptest.NewLogger(t)
	sess, _ := NewSession(priv, priv.PubKey(), salt, salt, logger)

	// send first 70 sequences (window size 64)
	for i := 0; i < 70; i++ {
		frame, _ := sess.EncodeRecord([]byte{byte(i)}, 0)
		if _, err := sess.DecodeRecord(frame); err != nil {
			t.Fatalf("unexpected decode fail at %d: %v", i, err)
		}
	}
	// re-send seq 1 (which is now out of window)
	oldFrame, _ := sess.EncodeRecord([]byte{0xFF}, 0)
	// manually overwrite seq number to 1
	binary.BigEndian.PutUint64(oldFrame[5:13], 1)
	if _, err := sess.DecodeRecord(oldFrame); err == nil {
		t.Fatal("expected old packet to be rejected")
	}
}
